### PR TITLE
fix: `Eventually()` missing `Should()` statement and sync error

### PR DESCRIPTION
**Pull request by jcanocan**
**What this PR does / why we need it**:
The linter enforces the usage of `Should()` statements when a `Eventually` check is used.
It adds the missing `Eventually()` statements `Should()` checks.

Also, `DeferCleanup` has been dropped in favor of `defer`. The [`resetToDefaultConfig()` ](https://github.com/kubevirt/kubevirt/blob/cb1b6e53540189d6664c4a8c126ab6e0a84ff8c4/tests/utils.go#L1842) is called before the test `DeferCleanup`, creating the following error:

```
  [FAILED] in [AfterEach] - tests/utils.go:1601 @ 01/29/24 15:32:28.576
  << Timeline
  [FAILED] Timed out after 10.001s.
  Unexpected error:
      <*errors.errorString | 0xc006203230>: 
      resource & config versions (5548 and 4736 respectively) are not as expected. component: "virt-handler", pod: "virt-handler-zdv7f" 
      {
          s: "resource & config versions (5548 and 4736 respectively) are not as expected. component: \"virt-handler\", pod: \"virt-handler-zdv7f\" ",
      }
  occurred
  In [AfterEach] at: tests/utils.go:1601 @ 01/29/24 15:32:28.576
  Full Stack Trace
    kubevirt.io/kubevirt/tests.resetToDefaultConfig()
        tests/utils.go:1601 +0x85
    kubevirt.io/kubevirt/tests.TestCleanup()
        tests/utils.go:113 +0x6f
    kubevirt.io/kubevirt/tests_test.glob..func21()
        tests/tests_suite_test.go:108 +0xf
```

This is because the `virt-handler` will not be ready (intentionally for the purposes of the test), and the `resetToDefaultConfig()` will force the `virt-handler` to reconcile, which will fail, but the `virt-handler` `resourceVersion` will be updated. Therefore, the Kubevirt object will be out of sync.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by xpivarc — APPROVED** [review]
/approve

**Comment by xpivarc** [comment]
/hold 
flakes

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1989]: @jcanocan Please remove this.

File: tests/vm_test.go
Comment (EdDev) [inline tests/vm_test.go:1977]: These kind of hacks are causing the whole tests suite to collapse on itself.

These is nothing safe that can be done to fix something like this and I hope to see this whole test removed.

But this is not really related to this fix. I just think the fix is to remove it.

File: tests/vm_test.go
Comment (fossedihelm) [inline tests/vm_test.go:1972]: How about?
```
Eventually(func(g Gomega) {
	g.Expect(getHandlerNodePod(virtClient, nodeName).Items[0]).To(HaveConditionFalse(k8sv1.PodReady))
}, 120*time.Second, time.Second).Should(Succeed())

DeferCleanup(func() {
	libpod.DeleteKubernetesApiBlackhole(getHandlerNodePod(virtClient, nodeName), componentName)
	Eventually(func(g Gomega) {
		g.Expect(getHandlerNodePod(virtClient, nodeName).Items[0]).To(HaveConditionTrue(k8sv1.PodReady))
	}, 120*time.Second, time.Second).Should(Succeed())
})
```

File: tests/vm_test.go
Comment (jcanocan) [reply tests/vm_test.go]: As far as I was able to test, this has been the key piece to stabilize the test. I've taken it from this other test: https://github.com/kubevirt/kubevirt/blob/7cfc6256f85e2853c9571f60e0b1087775cb820b/tests/virt_control_plane_test.go#L271-L279
Dropped the `FIXME` comment.

Comment (jcanocan) [reply tests/vm_test.go]: I will give the flakes line a couple of extra runs to be sure.

File: tests/vm_test.go
Comment (jcanocan) [reply tests/vm_test.go]: Dropped the usage of  `defer` for this purpose. I agree, it's not elegant. Thanks for the input.

File: tests/vm_test.go
Comment (fossedihelm) [inline tests/vm_test.go:1962]: `AfterEach` is enough here. Inner `AfterEach` are executed before the outer one

Comment (jcanocan) [reply tests/vm_test.go]: Done.

**Comment by fossedihelm** [comment]
/lgtm
Thanks!
@xpivarc would you like to take a further look before unhold?

File: tests/vm_test.go
Comment (xpivarc) [inline tests/vm_test.go:1970]: This is not flaky?

File: tests/vm_test.go
Comment (xpivarc) [reply tests/vm_test.go:1972]: Why did we revert the `defer` or `JustAfterEach`?

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: tests/vm_test.go
Comment (EdDev) [reply tests/vm_test.go]: I was actually referring to hacking the virt-handler network.

It is indeed nicer without the `defer`, but what actually is done here is very risky and IMO does not fit our tests.
If for some reason the pod state is not fixed (e.g. the route is not deleted), we are left with a broken cluster and following tests will flake in a hard to understand manner.

IMO we should not have such tests around, we cannot cover every scenario in a reasonable safe manner.
Our resources are limited, tests are long and flaky enough not to risk them further.
If someone thinks it is worth the resources and risk, it would be better to allocate a dedicate job for such tests.  I think reality dictates that we need to manage our e2e tests and focus on the main functionality and leave the edge scenarios to the community to detect (reporting bugs, running more tests on D/S projects and contributing back).

Comment (xpivarc) [reply tests/vm_test.go]: There is precedence for destructive tests, if we want to run them on separate lanes or not have them at all, the decision should be made in wider consensus + maintainers.

I am not sure what is better about the `AfterEach`, other than it is wrong afaik.

Note: I don't agree. This is not a conformance test and if anything this does fail loudly!

Comment (EdDev) [reply tests/vm_test.go]: Well, I disagreeing with the presence of cluster-level destructive tests and I reasoned about it in the prev message. The fact that they exist is not a sign of them being good for the overall project.

> This is not a conformance test and if anything this does fail loudly!

If tests like this fail, they either mess up all following tests or destabilize the cluster in ways we may not predict.

Comment (dominikholler) [reply tests/vm_test.go]: > There is precedence for destructive tests, if we want to run them on separate lanes or not have them at all, the decision should be made in wider consensus + maintainers.

> Well, I disagree with the presence of cluster-level destructive tests and I reasoned about it in the prev message. The fact that they exist is not a sign of them being good for the overall project.

> > This is not a conformance test and if anything this does fail loudly!

> If tests like this fail, they either mess up all following tests or destabilize the cluster in ways we may not predict.


We have no rush to get this PR merged and from my point of view these observations are worth to be discussed like @xpivarc proposed. What would be a good way of shifting this discussion away from the PR to a wider forum?

File: tests/vm_test.go
Comment (fossedihelm) [reply tests/vm_test.go:1972]: I don't get it.
The `defer` is something that we can avoid as solution to cleanup. The `JustAfterEach` is generally used for different scopes.
Here, the problem was the `DeferCleanup`. After some searches I found this https://github.com/onsi/ginkgo/issues/1284#issuecomment-1756314394
```
In reality Ginkgo runs the After* family followed by the Defer* family.
```
which explain the source of the issue.
(We should check our other `DeferCleanup` usages IMHO).
Can you explain your doubts?
Thanks

Comment (xpivarc) [reply tests/vm_test.go:1972]: @jcanocan also observed that an `AfterEach` is only run after the global `AfterEach` afaik.

File: tests/vm_test.go
Comment (brianmcarey) [reply tests/vm_test.go]: I will be drafting a proposal for the wider kubevirt community to move these kind of `destructive` and edge case tests to their own separate test lane and out of the gating lanes. We are currently testing everything in the gating lanes including severe edge cases which does not make sense and makes CI very costly. 

We will look at how to separate out these tests and where we want to run them.

File: tests/vm_test.go
Comment (jcanocan) [reply tests/vm_test.go:1972]: Thanks @fossedihelm for this nice piece of information. This explains all the issues faced.

As far as I was able to test, the nearest `AfterEach` (the one inside the `Context`) is always executed before the global `AfterEach`. However, we need a way to wait until the `handler` and `kubevirt` versions are in sync. That's why I've observed in the past the error mentioned in the description. Sorry for the confusion.

**Comment by jcanocan** [comment]
@xpivarc flakes failed, but no because this test. Please, retest it when you have taken a look.

**Comment by xpivarc** [comment]
/hold cancel 
@fossedihelm

**Review by fossedihelm — APPROVED** [review]
Thank you!

**Pull request merged**
